### PR TITLE
[#77] Support for VisitType.

### PR DESCRIPTION
**Pull request by reagan-meant**
Need help with 
1. overiding description using only name
2. Voiding concepts using only the name without uuid
3. overide description using uuid

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/initializer/Domain.java [hunk 1]
```diff
@@ -30,7 +30,8 @@
 	METADATA_SETS(26, "metadatasets"),
 	METADATA_SET_MEMBERS(27, "metadatasetmembers"),
 	METADATA_TERM_MAPPINGS(28, "metadatatermmappings"),
-	HTML_FORMS(29, "htmlforms");
+	HTML_FORMS(29, "htmlforms"),
+  VISIT_TYPES(30, "visittypes");
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/initializer/Domain.java:34]: Visit types can presumably be loaded very early, I guess this should work:
```java
VISIT_TYPES(3, "visittypes");
```

File: api/src/main/java/org/openmrs/module/initializer/api/visittypes/VisitTypeCsvParser.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import org.openmrs.module.initializer.Domain;
+import org.openmrs.module.initializer.api.BaseLineProcessor;
+import org.openmrs.module.initializer.api.CsvLine;
+import org.openmrs.module.initializer.api.CsvParser;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.stereotype.Component;
+
+@Component
+public class VisitTypeCsvParser extends CsvParser<VisitType, BaseLineProcessor<VisitType>> {
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/initializer/api/visittypes/VisitTypeCsvParser.java:18]: `VisitTypeCsvParser` → `VisitTypesCsvParser`

File: api/src/main/java/org/openmrs/module/initializer/api/visittypes/VisitTypeCsvParser.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+			List<VisitType> visitTypes = visitService.getVisitTypes(line.getName(true));
+			
+			for (VisitType v : visitTypes) {
+
+				if (v.getName() == line.getName(true)) {
+					visitType = v;
+				
+				}
+			}
+		}
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/initializer/api/visittypes/VisitTypeCsvParser.java:50]: This could be done in one line using streams.

File: api/src/main/java/org/openmrs/module/initializer/api/visittypes/VisitTypeLineProcessor.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+import org.openmrs.VisitType;
+import org.openmrs.api.VisitService;
+import org.openmrs.module.initializer.api.BaseLineProcessor;
+import org.openmrs.module.initializer.api.CsvLine;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.stereotype.Component;
+
+/**
+ * This is the first level line processor for visitTypes. It allows to parse and
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/initializer/api/visittypes/VisitTypeLineProcessor.java:12]: In plain English just call them as they are: "visit types" (not "visitTypes").

File: api/src/main/java/org/openmrs/module/initializer/api/visittypes/VisitTypeLineProcessor.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+ */
+@Component
+public class VisitTypeLineProcessor extends BaseLineProcessor<VisitType> {
+
+	/*
+	 * protected static String HEADER_NAME = "name";
+	 * 
+	 * protected static String HEADER_DESCRIPTION = "description";
+	 * 
+	 */
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/initializer/api/visittypes/VisitTypeLineProcessor.java:23]: This should go.

File: api/src/main/java/org/openmrs/module/initializer/api/visittypes/VisitTypeLineProcessor.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+	@Autowired
+	public VisitTypeLineProcessor(@Qualifier("visitService") VisitService visitService) {
+		this.visitService = visitService;
+	}
+
+	@Override
+	public VisitType fill(VisitType visitType, CsvLine line) throws IllegalArgumentException {
+
+		visitType.setName(line.getName(true));
+		visitType.setDescription(line.getString(HEADER_DESC, line.getName(true) + " Visit"));
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/initializer/api/visittypes/VisitTypeLineProcessor.java:35]: What's going on with the description here?

File: api/src/test/java/org/openmrs/module/initializer/api/VisitTypeLoaderIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+	public void setup() {
+
+		// VisitType added directly
+		{
+			VisitType v = new VisitType();
+			v.setUuid("2bcf7212-d218-4572-8893-25c4pob71934");
+			v.setName("Malnutrition");
+			v.setDescription("Malnutrition Visit");
+			v = vs.saveVisitType(v);
+		}
```
Comment (mks-d) [inline api/src/test/java/org/openmrs/module/initializer/api/VisitTypeLoaderIntegrationTest.java:43]: Add it in [here](https://github.com/mekomsolutions/openmrs-module-initializer/blob/aca9ad4d4356707b0ea717ccfb219546dd0bb7d6/api/src/test/resources/testdata/test-metadata.xml), not through code.

**Review by mks-d — CHANGES_REQUESTED** [review]
You will also need to add a new entry in [`LoadersIntegrationTest`](https://github.com/mekomsolutions/openmrs-module-initializer/blob/aca9ad4d4356707b0ea717ccfb219546dd0bb7d6/api-2.3/src/test/java/org/openmrs/module/initializer/api/loaders/LoadersIntegrationTest.java#L48).

File: api/src/main/java/org/openmrs/module/initializer/api/visittypes/VisitTypeLineProcessor.java (hunk 6 above)
Comment (reagan-meant) [reply api/src/main/java/org/openmrs/module/initializer/api/visittypes/VisitTypeLineProcessor.java:35]: This is to set a description in case one is not set in the csv

File: api/src/main/java/org/openmrs/module/initializer/api/visittypes/VisitTypeLineProcessor.java (hunk 4 above)
Comment (reagan-meant) [reply api/src/main/java/org/openmrs/module/initializer/api/visittypes/VisitTypeLineProcessor.java:12]: Done

File: api/src/main/java/org/openmrs/module/initializer/Domain.java (hunk 1 above)
Comment (reagan-meant) [reply api/src/main/java/org/openmrs/module/initializer/Domain.java]: Done

File: api/src/main/java/org/openmrs/module/initializer/api/visittypes/VisitTypeCsvParser.java (hunk 2 above)
Comment (reagan-meant) [reply api/src/main/java/org/openmrs/module/initializer/api/visittypes/VisitTypeCsvParser.java]: Done

File: api/src/main/java/org/openmrs/module/initializer/api/visittypes/VisitTypeCsvParser.java (hunk 3 above)
Comment (reagan-meant) [reply api/src/main/java/org/openmrs/module/initializer/api/visittypes/VisitTypeCsvParser.java]: Done

File: api/src/main/java/org/openmrs/module/initializer/api/visittypes/VisitTypeLineProcessor.java (hunk 5 above)
Comment (reagan-meant) [reply api/src/main/java/org/openmrs/module/initializer/api/visittypes/VisitTypeLineProcessor.java]: Done

File: api/src/test/java/org/openmrs/module/initializer/api/VisitTypeLoaderIntegrationTest.java (hunk 7 above)
Comment (reagan-meant) [reply api/src/test/java/org/openmrs/module/initializer/api/VisitTypeLoaderIntegrationTest.java]: Done

File: api-2.3/src/test/java/org/openmrs/module/initializer/api/loaders/LoadersIntegrationTest.java [hunk 8]
```diff
@@ -159,6 +163,11 @@ public void getLoaders_shouldBeUnivoquelyOrdered() {
 		count++;
 		Assert.assertThat(loader.getOrder(), greaterThan(previousLoader.getOrder()));
 		
+		previousLoader = loader;
+		loader = visitTypeLoader;
```
Comment (mks-d) [inline api-2.3/src/test/java/org/openmrs/module/initializer/api/loaders/LoadersIntegrationTest.java:167]: `visitTypeLoader` → `visitTypesLoader`

File: api/src/main/java/org/openmrs/module/initializer/api/visittypes/VisitTypeLineProcessor.java (hunk 6 above)
Comment (mks-d) [reply api/src/main/java/org/openmrs/module/initializer/api/visittypes/VisitTypeLineProcessor.java:29]: Just cater for a description column and do like any other domain would do, encounter types or locations for instance.

File: api/src/main/java/org/openmrs/module/initializer/api/visittypes/VisitTypeLoader.java [hunk 9]
```diff
@@ -7,7 +7,7 @@
 
 @Component
 public class VisitTypeLoader extends BaseCsvLoader<VisitType, VisitTypesCsvParser> {
-
+
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/initializer/api/visittypes/VisitTypeLoader.java:10]: I forgot that one sorry, `VisitTypeLoader` → `VisitTypesLoader`.

The rule is:
* The loader applies for _all_ visit types.
* The parser applies for _all_ visit types.
* The line processor applies for _one_ visit type at a time.

File: api/src/main/java/org/openmrs/module/initializer/api/visittypes/VisitTypesCsvParser.java [hunk 10]
```diff
@@ -40,7 +40,8 @@ public VisitType bootstrap(CsvLine line) throws IllegalArgumentException {
 		VisitType visitType = visitService.getVisitTypeByUuid(uuid);
 		if (visitType == null) {
 			
-			visitType = visitService.getVisitTypes(line.getName(true)).stream().filter(vt -> vt.getName() == line.getName(true)).findFirst().map(vt-> vt).orElse(null);
+			visitType = visitService.getVisitTypes(line.getName(true)).stream()
+			        .filter(vt -> vt.getName() == line.getName(true)).findFirst().map(vt -> vt).orElse(null);
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/initializer/api/visittypes/VisitTypesCsvParser.java:44]: Good!

File: api/src/test/java/org/openmrs/module/initializer/api/VisitTypeLoaderIntegrationTest.java [hunk 11]
```diff
@@ -69,13 +69,13 @@ public void load_shouldLoadVisitTypeAccordingToCsvFiles() throws Exception {
 			// Need to make this pass
 			// Assert.assertEquals("TB Description", v.getDescription());
 		}
-
+		
 		// Test to show that Iniz can retire VisitType using uuid only
 		{
 			VisitType v = vs.getVisitTypeByUuid("l3cf7209-d218-4572-8893-25c4b5b71934");
-			 Assert.assertEquals(true,v.isRetired());
+			Assert.assertEquals(true, v.isRetired());
```
Comment (mks-d) [inline api/src/test/java/org/openmrs/module/initializer/api/VisitTypeLoaderIntegrationTest.java:76]: Name them `vt` rather than `v`.

**Review by mks-d — CHANGES_REQUESTED** [review]

File: readme/visittypes.md [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+The **visittypes** subfolder contains CSV import files for saving visittypes in bulk. This is a possible example of its content:
+```bash
+visittypes/
+  ├──visittypes.csv
+  └── ...
+```
+There is currently only one format for the visittype CSV line, here are the possible headers with a sample data set:
+
+|<sub>Uuid</sub> |<sub>Void/Retire</sub> |<sub>Name</sub> | <sub>Description</sub> | <sub>_order:1000</sub> |
+| - | - | - | - | - | - |
```
Comment (mks-d) [inline readme/visittypes.md:11]: This doesn't display properly because you set 6 columns instead of 5 here:
```md
| - | - | - | - | - |
```

File: readme/visittypes.md [hunk 13]
```diff
@@ -0,0 +1,27 @@
+## Domain 'visittypes'
+The **visittypes** subfolder contains CSV import files for saving visittypes in bulk. This is a possible example of its content:
+```bash
+visittypes/
+  ├──visittypes.csv
+  └── ...
+```
+There is currently only one format for the visittype CSV line, here are the possible headers with a sample data set:
```
Comment (mks-d) [inline readme/visittypes.md:8]: Plain English: `visittype` → `visit type`.

File: readme/visittypes.md [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+
+
+Let's review some important headers.
+
+###### Header `Name`
+This is _not_ a localized header.
+<br/>The  Name is a secondary identifier to access a visittype, it will be used to attempt fetching the visit type if no UUID is provided.
+
+###### Header `Description`
+A Description is used to give more information about the Visit type.
```
Comment (mks-d) [inline readme/visittypes.md:24]: Plain English: `Visit type` → `visit type`.

File: readme/visittypes.md [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+|<sub>9o805171-afcb-444f-84ec-3f7df078628f</sub>|<sub></sub>| <sub>General </sub> | <sub>A General Visit</sub> |<sub></sub>|
+
+Headers that start with an underscore such as `_order:1000` are metadata headers. The values in the columns under those headers are never read by the CSV parser.
+
+
+Let's review some important headers.
+
+###### Header `Name`
+This is _not_ a localized header.
+<br/>The  Name is a secondary identifier to access a visittype, it will be used to attempt fetching the visit type if no UUID is provided.
```
Comment (mks-d) [inline readme/visittypes.md:21]: Plain English: `visittype` → `visit type`.

File: readme/visittypes.md [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+|<sub>Uuid</sub> |<sub>Void/Retire</sub> |<sub>Name</sub> | <sub>Description</sub> | <sub>_order:1000</sub> |
+| - | - | - | - | - | - |
+|<sub>9o805171-afcb-444f-84ec-3f7df078628f</sub>|<sub></sub>| <sub>General </sub> | <sub>A General Visit</sub> |<sub></sub>|
+
+Headers that start with an underscore such as `_order:1000` are metadata headers. The values in the columns under those headers are never read by the CSV parser.
+
+
+Let's review some important headers.
+
+###### Header `Name`
```
Comment (mks-d) [inline readme/visittypes.md:19]: Please specify mandatory fields like here: https://github.com/mekomsolutions/openmrs-module-initializer/blob/master/readme/autogenerationoptions.md#header-identifier-type-mandatory

File: api/src/test/resources/testAppDataDir/configuration/visittypes/visittypes.csv [hunk 17]
```diff
@@ -2,7 +2,8 @@ Uuid,Void/Retire,Name,Description,_version:1
 2bcf7212-d218-4572-88o9-25c4b5b71934,,VIH,HIV Description,
 abcf7209-d218-4572-8893-25c4b5b71934,,General,,
 l3cf7209-d218-4572-8893-25c4b5b71934,,ANTENATAL,,
-xyzf7209-d218-4572-8893-xyz4b5b71934,,RANDOM,,
+xyzf7209-d218-4572-8893-xyz4b5b71934,,RANDOM,RANDOM DESC,
```
Comment (mks-d) [inline api/src/test/resources/testAppDataDir/configuration/visittypes/visittypes.csv:5]: Please provide a valid UUID and values that look like real world examples.

**Review by mks-d — CHANGES_REQUESTED** [review]

File: api/src/test/java/org/openmrs/module/initializer/api/VisitTypesLoaderIntegrationTest.java [hunk 18]
```diff
@@ -56,8 +56,8 @@ public void load_shouldLoadVisitTypeAccordingToCsvFiles() throws Exception {
 		}
 		// Test to confirm Description is changed when uuid is specified in csv
 		{
-			VisitType vt = vs.getVisitTypeByUuid("xyzf7209-d218-4572-8893-xyz4b5b71934");
-			Assert.assertEquals("RANDOM DESC CHANGED", vt.getDescription());
+			VisitType vt = vs.getVisitTypeByUuid("287463d3-2233-4c69-9851-5841a1f5e109");
+			Assert.assertEquals("Out Patient Visit", vt.getDescription());
```
Comment (mks-d) [inline api/src/test/java/org/openmrs/module/initializer/api/VisitTypesLoaderIntegrationTest.java:60]: "Outpatient" writes in one word, or you could use "OPD Visit".

**Review by mks-d — CHANGES_REQUESTED** [review]
Almost there.

File: api/src/test/java/org/openmrs/module/initializer/api/VisitTypesLoaderIntegrationTest.java [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+		executeDataSet("testdata/test-metadata.xml");
+		
+		// Replay
+		loader.load();
+		
+		// Test to find visitTypes with the phrase TB...
+		{
+			List<VisitType> vts = vs.getVisitTypes("TB");
+			Assert.assertNotNull(vts);
+			Assert.assertEquals(1, vts.size());
```
Comment (mks-d) [inline api/src/test/java/org/openmrs/module/initializer/api/VisitTypesLoaderIntegrationTest.java:43]: You should also assert that the name is the one expected for the one visit type that was fetched.

File: api/src/test/java/org/openmrs/module/initializer/api/VisitTypesLoaderIntegrationTest.java [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+		{
+			VisitType vt = vs.getVisitTypeByUuid("2bcf7212-d218-4572-8893-25c4pob71934");
+			Assert.assertNotNull(vt);
+			Assert.assertEquals("Malnutrition Visit", vt.getDescription());
+		}
+		
+		// Test to confirm Description is not added when none is given in csv
+		{
+			VisitType vt = vs.getVisitTypeByUuid("abcf7209-d218-4572-8893-25c4b5b71934");
+			Assert.assertNull(vt.getDescription());
```
Comment (mks-d) [inline api/src/test/java/org/openmrs/module/initializer/api/VisitTypesLoaderIntegrationTest.java:55]: Assert the name, that's the main thing to assert after all.
I mean visit types are just three things:
1. A UUID.
2. A name.
3. A description.

In your tests, at minima, when you fetch them by name, assert the UUID, when you fetch by UUID assert the name. That might be different in case of retirements of course.

**Review by mks-d — CHANGES_REQUESTED** [review]

File: api/src/test/java/org/openmrs/module/initializer/api/VisitTypesLoaderIntegrationTest.java (hunk 20 above)
Comment (reagan-meant) [reply api/src/test/java/org/openmrs/module/initializer/api/VisitTypesLoaderIntegrationTest.java:58]: Done

File: api/src/test/java/org/openmrs/module/initializer/api/VisitTypesLoaderIntegrationTest.java (hunk 19 above)
Comment (reagan-meant) [reply api/src/test/java/org/openmrs/module/initializer/api/VisitTypesLoaderIntegrationTest.java:43]: Done

File: api/src/test/java/org/openmrs/module/initializer/api/VisitTypesLoaderIntegrationTest.java [hunk 21]
```diff
@@ -0,0 +1,87 @@
...
+			VisitType vt = vs.getVisitTypeByUuid("abcf7209-d218-4572-8893-25c4b5b71934");
+			Assert.assertEquals("General", vt.getName());
+			Assert.assertNull(vt.getDescription());
+		}
+		// Test to confirm Description is changed when uuid is specified in csv
+		{
+			VisitType vt = vs.getVisitTypeByUuid("287463d3-2233-4c69-9851-5841a1f5e109");
+			Assert.assertEquals("OPD", vt.getName());
+			Assert.assertEquals("OPD Visit", vt.getDescription());
+		}
```
Comment (mks-d) [inline api/src/test/java/org/openmrs/module/initializer/api/VisitTypesLoaderIntegrationTest.java:65]: This is the CSV:
```csv
287463d3-2233-4c69-9851-5841a1f5e109,,OPD,OPD,
```
This shows that the description is in fact `"OPD"`, so how come you assert it to be `"OPD Visit"`?

Comment (reagan-meant) [reply api/src/test/java/org/openmrs/module/initializer/api/VisitTypesLoaderIntegrationTest.java:65]: I am overriding it [here](https://github.com/mekomsolutions/openmrs-module-initializer/pull/89/files#diff-42dd0f50285fb1e647f4943649294052200e36f560b6616998e89d1fb5189dbdR9)

Comment (mks-d) [reply api/src/test/java/org/openmrs/module/initializer/api/VisitTypesLoaderIntegrationTest.java:65]: I see.
Well if you want to make a test that shows that a description can be edited via CSV using the name as a pivot identifier, then you would need to do this:
1. Pre-create the visit type through the test dataset.
2. Have only the final version of the visit type in the CSV file.
3. Assert that what you get in the end is the resulting visit type as per the CSV file.

Comment (mks-d) [reply api/src/test/java/org/openmrs/module/initializer/api/VisitTypesLoaderIntegrationTest.java:65]: I see.
Well if you want to make a test that shows that a description can be edited via CSV using the name as a pivot identifier, then you would need to do this:
1. Pre-create the visit type through the test dataset.
2. Have only the final version of the visit type in the CSV file.
3. Assert that what you get in the end is the resulting visit type as per the CSV file.

**Review by mks-d — CHANGES_REQUESTED** [review]

**Comment by rbuisson** [comment]
Hi @mks-d @reagan-meant , is this ready for merge ?

**Review by mks-d — APPROVED** [review]
